sbi/src/api/tsm.rs

```diff
@@ -0,0 +1,280 @@
+    });
+    // Safety: The passed-in pages are unmapped and we do not access them again until they're
+    // reclaimed.
+    ecall_send(&msg)?;
+
+    // Fence the pages we just converted.
+    initiate_fence()
+}
+
+/// Reclaims pages that were previously converted to confidential memory with `convert_pages`.
+pub fn reclaim_pages(addr: u64, num_pages: u64) -> Result<()> {
+    let msg = SbiMessage::Tee(TsmReclaimPages {
+        page_addr: addr,
+        page_type: TsmPageType::Page4k,
+        num_pages,
+    });
+    // Safety: The referenced pages are made accessible again, which is safe since we haven't
+    // done anything with them since they were converted.
+    unsafe { ecall_send(&msg).map(|_| ()) }
+}
+
+/// Creates a new TVM.
+///
+/// # Params:
+///
+/// - tvm_page_directory_addr: The base physical address of the 16kB confidential memory region that
+/// should be used for the TVM's page directory. Must be 16kB-aligned.
+///
+/// - tvm_state_addr: The base physical address of the confidential memory region to be used to hold
+/// the TVM's global state. Must be page-aligned and `TsmInfo::tvm_state_pages` pages in length.
+///
+/// - tvm_num_vcpus: The maximum number of vCPUs that will be created for this TVM. Must be less
+/// than or equal to `TsmInfo::tvm_max_vcpus`.
+///
+/// - tvm_vcpu_addr: The base physical address of the confidential memory region to be used to hold
+/// the TVM's vCPU state. Must be page-aligned and `TsmInfo::tvm_bytes_per_vcpu` * `tvm_num_vcpus`
+/// bytes in length, rounded up to the nearest multiple of 4kB.
+pub fn tvm_create(
+    tvm_page_directory_addr: u64,
+    tvm_state_addr: u64,
+    tvm_num_vcpus: u64,
+    tvm_vcpu_addr: u64,
+) -> Result<u64> {
+    let tvm_create_params = TvmCreateParams {
+        tvm_page_directory_addr,
+        tvm_state_addr,
+        tvm_num_vcpus,
+        tvm_vcpu_addr,
+    };
+    let msg = SbiMessage::Tee(TvmCreate {
+        params_addr: (&tvm_create_params as *const TvmCreateParams) as u64,
+        len: core::mem::size_of::<TvmCreateParams>() as u64,
+    });
+    // Safety: creating a TVM will only touch pages that have already been converted to confidential
+    // memory, so it can't affect memory safety as the host doesn't have access to those pages.
+    let vmid = unsafe { ecall_send(&msg)? };
+    Ok(vmid)
+}
+
+/// Finalizes the given TVM
+pub fn tvm_finalize(vmid: u64) -> Result<()> {
+    let msg = SbiMessage::Tee(Finalize { guest_id: vmid });
+    // Safety: `Finalize` doesn't touch memory.
+    unsafe { ecall_send(&msg).map(|_| ()) }
+}
+
+/// Destroys a TVM created with `tvm_create`.
+pub fn tvm_destroy(vmid: u64) -> Result<()> {
+    let msg = SbiMessage::Tee(TvmDestroy { guest_id: vmid });
+    // Safety: destroying a VM doesn't write to memory that's accessible from the host.
+    unsafe { ecall_send(&msg).map(|_| ()) }
+}
+
+/// Runs the given vcpu of the specified TVM.
+pub fn tvm_run(vmid: u64, vcpu_id: u64) -> Result<u64> {
```

Comment:
Maybe convert the u64 to a `TvmCpuExitCode`?

Reply:
good idea, done